sectionfilter.c

```diff
@@ -270,54 +271,68 @@ cSatipSectionFilterHandler::~cSatipSectionFilterHandler()
+  return true;
+}
+
 void cSatipSectionFilterHandler::Action(void)
 {
   debug1("%s Entering [device %d]", __PRETTY_FUNCTION__, deviceIndexM);
-  bool processed = false;
   // Do the thread loop
+  uchar *p = NULL;
   while (Running()) {
+        int len = 0;
+        // Process all pending TS packets
+        while ((p  = ringBufferM->Get(len)) != NULL) {
+              if (p && (len >= TS_SIZE)) {
+                 if (*p != TS_SYNC_BYTE) {
+                    for (int i = 1; i < len; ++i) {
+                        if (p[i] == TS_SYNC_BYTE) {
+                           len = i;
+                           break;
+                           }
+                        }
+                    ringBufferM->Del(len);
+                    debug1("%s Skipped %d bytes to sync on TS packet [device %d]", __PRETTY_FUNCTION__, len, deviceIndexM);
+                    continue;
+                    }
+                    // Process TS packet through all filters
+                    mutexM.Lock();
+                    for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
+                        if (filtersM[i])
+                           filtersM[i]->Process(p);
+                        }
+                    mutexM.Unlock();
+                    ringBufferM->Del(TS_SIZE);
+                    }
+                }
+
         // Send demuxed section packets through all filters
-        bool retry = false;
         mutexM.Lock();
-        for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
-            if (filtersM[i] && filtersM[i]->Send())
-               retry = true;
-            }
+        while (Send()) ;
         mutexM.Unlock();
-        if (retry)
-           continue;
-        // Read one TS packet
-        if (ringBufferM) {
-           int len = 0;
-           if (processed) {
-              ringBufferM->Del(TS_SIZE);
-              processed = false;
-              }
-           uchar *p = ringBufferM->Get(len);
-           if (p && (len >= TS_SIZE)) {
-              if (*p != TS_SYNC_BYTE) {
-                 for (int i = 1; i < len; ++i) {
-                     if (p[i] == TS_SYNC_BYTE) {
-                        len = i;
-                        break;
-                        }
-                     }
-                 ringBufferM->Del(len);
-                 debug1("%s Skipped %d bytes to sync on TS packet [device %d]", __PRETTY_FUNCTION__, len, deviceIndexM);
-                 continue;
-                 }
-              // Process TS packet through all filters
-              mutexM.Lock();
-              for (unsigned int i = 0; i < eMaxSecFilterCount; ++i) {
-                  if (filtersM[i])
-                     filtersM[i]->Process(p);
-                  }
-              mutexM.Unlock();
-              processed = true;
-              continue;
-              }
-           }
-        cCondWait::SleepMs(10); // to avoid busy loop and reduce cpu load
+        //cCondWait::SleepMs(10); // to avoid busy loop and reduce cpu load
```

Comment:
We should sleep - at least if there weren't anything to send.

Reply:
AFAIK, cRingBuffer::Get() blocks if the buffer is empty. I think sleeping here is unneccessary. I do not see any difference in CPU usage with of without the sleep. But maybe I'm wrong.

Reply:
Oh, I forgot that there are 100ms timeouts defined in the constructor for the ring buffer. Please, remove the dead code.